### PR TITLE
Add information about using assertion error and runtime errors.

### PR DESCRIPTION
**Pull request by adiroiban**
Scope
=====

This add some info about how to use assertion error and how not to use the assert statement.

How to try and test the changes
===============================

reviewers: @brunogola @alibotean 

check that changes make sense.

feel free to send your feedback

### CODE REVIEW

**Comment by adiroiban** [comment]
needs-review

File: content/generic.rst [hunk 1]
```diff
@@ -260,6 +260,28 @@ Each ID should be raised from a single place.
...
+disabled.
+
+Don't raise `AssertionError` outside of the test code.
+Use `RuntimeError` or a more specific exception.
+The `AssertionError` should never be handled nor in production nor in
+testing code.
+Using `self.assertRaises(AssertionError)` is a form of handling an
+exception.
+
+Don't raise `RuntimeError` outside for the production code.
```
Comment (brunogola) [inline content/generic.rst]: i believe you meant outside **of** ?

**Comment by brunogola** [comment]
one minor comment

changes-approved

File: content/generic.rst [hunk 2]
```diff
@@ -260,6 +260,28 @@ Each ID should be raised from a single place.
...
 
+Don't use the `assert` statement in code but rather raise an explicit error.
+`assert` statement optimization is useless as it was scientifically proven
+that some bugs only show up when a customer uses the machine and we want
+those exceptions to be raised in production and not be accidentally
+disabled.
+
+Don't raise `AssertionError` outside of the test code.
+Use `RuntimeError` or a more specific exception.
+The `AssertionError` should never be handled nor in production nor in
```
Comment (alibotean) [inline content/generic.rst]: The `AssertionError` should never be handled; be in production or testing code.

File: content/generic.rst [hunk 3]
```diff
@@ -260,6 +260,28 @@ Each ID should be raised from a single place.
...
+testing code.
+Using `self.assertRaises(AssertionError)` is a form of handling an
```
Comment (alibotean) [inline content/generic.rst]: I would place this and the above statement in their own paragraph; taking this into account I find it hard to understand when will we use them at all:

```
Don't raise `AssertionError` outside of the test code. [...] The `AssertionError` should never be handled [...] in testing code. Using `self.assertRaises(AssertionError)` is a form of handling an exception.
```

File: content/generic.rst [hunk 4]
```diff
@@ -260,6 +260,28 @@ Each ID should be raised from a single place.
...
+exception.
+
+Don't raise `RuntimeError` outside for the production code.
+Use `AssertionError` when implementing doubles, stub or mock implementation
+to support the testing.
+
+All raised assertion should have a descriptive message.
```
Comment (alibotean) [inline content/generic.rst]: assertions

**Comment by alibotean** [comment]
Just some minor comments.

changes-approved

**Comment by adiroiban** [comment]
thanks for the review. for typos and other stuff feel free to commit the changes

**Pull request merged**